### DISCUSSION TITLE
searchable metadata throws out error. Showing HTTP 403 error in logs.

### DISCUSSION THREAD

**Discussion by apagapulan-cwt** [thoughtspot/cs_tools#166] [General]
Hi Team,

When we try to run the cs tools searchable metadata, we get unhandled error.
Upon checking the logs, i saw these unusual lines of error messages. 

{"type":"Forbidden","description":"The client does not have access rights to the content; that is,it is unauthorized, so the server is refusing to give the requested resource. Unlike 401, the client\u0027s identity is known to the server."}

May I know what went wrong? I was able to run some reports before we upgrade to 10.0 last August 12. 

Regards,
Archelle

**Comment by boonhapus** [comment]
@apagapulan-cwt are you using an admin account? Can you share a log file with the error in it, I'm wondering if this happens to all calls, or if it's for any particular object.

**Comment by apagapulan-cwt** [comment]
Hello, yes i am using admin account. See attached logs.

[2024-08-14T07_27_12_2.txt](https://github.com/user-attachments/files/16622293/2024-08-14T07_27_12_2.txt)

**Comment by apagapulan-cwt** [comment]
This what it looks like in my screen

C:\Users\UserId\Documents\cs_tools>cs_tools tools searchable metadata --org 0 --syncer csv://C:\Users\UserId\Documents\cs_tools\metadata_prod.toml --config prod
[14:44:52] INFO     Attempting Basic authentication                                                                thoughtspot.py:213
[14:44:55] INFO     Initializing syncer: <class 'cs_tools_csv_syncer.CSV'>                                               syncer.py:64
                                                      Searchable Data Gatherer

                              Org        User        Group        Tag       Object      Column     Dependent    Access
           ──────────────────────────────────────────────────────────────────────────────────────────────────────────────
             ThoughtSpot                  🔥          🍿          🍿          🍿          🍿          🍿          🍿

                                                           fetching data
╭──────────────────────────────────────────────── Traceback (most recent call last) ────────────────────────────────────────────────╮
│ C:\Users\UserId\AppData\Roaming\cs_tools\.cs_tools\Lib\site-packages\httpx\_transports\default.py:69 in map_httpcore_exceptions  │
│                                                                                                                                   │
│    66 @contextlib.contextmanager                                                                                                  │
│    67 def map_httpcore_exceptions() -> typing.Iterator[None]:                                                                     │
│    68 │   try:                                                                                                                    │
│ ❱  69 │   │   yield                                                                                                               │
│    70 │   except Exception as exc:                                                                                                │
│    71 │   │   mapped_exc = None                                                                                                   │
│    72                                                                                                                             │
│                                                                                                                                   │
│ C:\Users\UserId\AppData\Roaming\cs_tools\.cs_tools\Lib\site-packages\httpx\_transports\default.py:233 in handle_request          │
│                                                                                                                                   │
│   230 │   │   │   extensions=request.extensions,                                                                                  │
│   231 │   │   )                                                                                                                   │
│   232 │   │   with map_httpcore_exceptions():                                                                                     │
│ ❱ 233 │   │   │   resp = self._pool.handle_request(req)                                                                           │
│   234 │   │                                                                                                                       │
│   235 │   │   assert isinstance(resp.stream, typing.Iterable)                                                                     │
│   236                                                                                                                             │
│                                                                                                                                   │
│ C:\Users\UserId\AppData\Roaming\cs_tools\.cs_tools\Lib\site-packages\httpcore\_sync\connection_pool.py:216 in handle_request     │
│                                                                                                                                   │
│   213 │   │   │   │   closing = self._assign_requests_to_connections()                                                            │
│   214 │   │   │                                                                                                                   │
│   215 │   │   │   self._close_connections(closing)                                                                                │
│ ❱ 216 │   │   │   raise exc from None                                                                                             │
│   217 │   │                                                                                                                       │
│   218 │   │   # Return the response. Note that in this case we still have to manage                                               │
│   219 │   │   # the point at which the response is closed.                                                                        │
│                                                                                                                                   │
│ C:\Users\UserId\AppData\Roaming\cs_tools\.cs_tools\Lib\site-packages\httpcore\_sync\connection_pool.py:196 in handle_request     │
│                                                                                                                                   │
│   193 │   │   │   │                                                                                                               │
│   194 │   │   │   │   try:                                                                                                        │
│   195 │   │   │   │   │   # Send the request on the assigned connection.                                                          │
│ ❱ 196 │   │   │   │   │   response = connection.handle_request(                                                                   │
│   197 │   │   │   │   │   │   pool_request.request                                                                                │
│   198 │   │   │   │   │   )                                                                                                       │
│   199 │   │   │   │   except ConnectionNotAvailable:                                                                              │
│                                                                                                                                   │
│ C:\Users\UserId\AppData\Roaming\cs_tools\.cs_tools\Lib\site-packages\httpcore\_sync\connection.py:101 in handle_request          │
│                                                                                                                                   │
│    98 │   │   │   self._connect_failed = True                                                                                     │
│    99 │   │   │   raise exc                                                                                                       │
│   100 │   │                                                                                                                       │
│ ❱ 101 │   │   return self._connection.handle_request(request)                                                                     │
│   102 │                                                                                                                           │
│   103 │   def _connect(self, request: Request) -> NetworkStream:                                                                  │
│   104 │   │   timeouts = request.extensions.get("timeout", {})                                                                    │
│                                                                                                                                   │
│ C:\Users\UserId\AppData\Roaming\cs_tools\.cs_tools\Lib\site-packages\httpcore\_sync\http11.py:143 in handle_request              │
│                                                                                                                                   │
│   140 │   │   │   with ShieldCancellation():                                                                                      │
│   141 │   │   │   │   with Trace("response_closed", logger, request) as trace:                                                    │
│   142 │   │   │   │   │   self._response_closed()                                                                                 │
│ ❱ 143 │   │   │   raise exc                                                                                                       │
│   144 │                                                                                                                           │
│   145 │   # Sending the request...                                                                                                │
│   146                                                                                                                             │
│                                                                                                                                   │
│ C:\Users\UserId\AppData\Roaming\cs_tools\.cs_tools\Lib\site-packages\httpcore\_sync\http11.py:113 in handle_request              │
│                                                                                                                                   │
│   110 │   │   │   │   │   reason_phrase,                                                                                          │
│   111 │   │   │   │   │   headers,                                                                                                │
│   112 │   │   │   │   │   trailing_data,                                                                                          │
│ ❱ 113 │   │   │   │   ) = self._receive_response_headers(**kwargs)                                                                │
│   114 │   │   │   │   trace.return_value = (                                                                                      │
│   115 │   │   │   │   │   http_version,                                                                                           │
│   116 │   │   │   │   │   status,                                                                                                 │
│                                                                                                                                   │
│ C:\Users\UserId\AppData\Roaming\cs_tools\.cs_tools\Lib\site-packages\httpcore\_sync\http11.py:186 in _receive_response_headers   │
│                                                                                                                                   │
│   183 │   │   timeout = timeouts.get("read", None)                                                                                │
│   184 │   │                                                                                                                       │
│   185 │   │   while True:                                                                                                         │
│ ❱ 186 │   │   │   event = self._receive_event(timeout=timeout)                                                                    │
│   187 │   │   │   if isinstance(event, h11.Response):                                                                             │
│   188 │   │   │   │   break                                                                                                       │
│   189 │   │   │   if (                                                                                                            │
│                                                                                                                                   │
│ C:\Users\UserId\AppData\Roaming\cs_tools\.cs_tools\Lib\site-packages\httpcore\_sync\http11.py:238 in _receive_event              │
│                                                                                                                                   │
│   235 │   │   │   │   # it as a ConnectError.                                                                                     │
│   236 │   │   │   │   if data == b"" and self._h11_state.their_state == h11.SEND_RESPONSE:                                        │
│   237 │   │   │   │   │   msg = "Server disconnected without sending a response."                                                 │
│ ❱ 238 │   │   │   │   │   raise RemoteProtocolError(msg)                                                                          │
│   239 │   │   │   │                                                                                                               │
│   240 │   │   │   │   self._h11_state.receive_data(data)                                                                          │
│   241 │   │   │   else:                                                                                                           │
╰───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
RemoteProtocolError: Server disconnected without sending a response.

The above exception was the direct cause of the following exception:

╭──────────────────────────────────────────────── Traceback (most recent call last) ────────────────────────────────────────────────╮
│ C:\Users\UserId\AppData\Roaming\cs_tools\.cs_tools\Lib\site-packages\cs_tools\cli\tools\searchable\app.py:456 in metadata        │
│                                                                                                                                   │
│   453 │   │   │   row[2] = ":fire:"                                                                                               │
│   454 │   │   │   # user                                                                                                          │
│   455 │   │   │   # group_membership                                                                                              │
│ ❱ 456 │   │   │   r = ts.api.v1.user_read()                                                                                       │
│   457 │   │   │   members = {m["user_guid"] for m in temp_sync.load(models.User.__tablename__)                                    │
│   458 │   │   │   temp_sync.dump(                                                                                                 │
│   459 │   │   │   │   models.User.__tablename__, data=transform.to_user(r.json(), cluster=clus                                    │
│                                                                                                                                   │
│ C:\Users\UserId\AppData\Roaming\cs_tools\.cs_tools\Lib\site-packages\cs_tools\api\_rest_api_v1.py:164 in user_read               │
│                                                                                                                                   │
│   161 │   │   username: Union[str, _utils.UndefinedType] = _utils.UNDEFINED,                                                      │
│   162 │   ) -> httpx.Response:                                                                                                    │
│   163 │   │   p = {"userid": user_guid, "name": username}                                                                         │
│ ❱ 164 │   │   r = self.request("GET", "callosum/v1/tspublic/v1/user", params=p)                                                   │
│   165 │   │   return r                                                                                                            │
│   166 │                                                                                                                           │
│   167 │   def user_update(                                                                                                        │
│                                                                                                                                   │
│ C:\Users\UserId\AppData\Roaming\cs_tools\.cs_tools\Lib\site-packages\cs_tools\api\_rest_api_v1.py:58 in request                  │
│                                                                                                                                   │
│    55 │   │   """Pre-process the request to remove undefined parameters."""                                                       │
│    56 │   │   request_kw = _utils.scrub_undefined_sentinel(request_kw, null=_utils.UNDEFINED)                                     │
│    57 │   │   request_method = getattr(self._api_client, method.lower())                                                          │
│ ❱  58 │   │   return request_method(endpoint, **request_kw)                                                                       │
│    59 │                                                                                                                           │
│    60 │   # ==================================================================================                                    │
│    61 │   # SESSION     ::  https://developers.thoughtspot.com/docs/?pageid=rest-api-reference                                    │
│                                                                                                                                   │
│ C:\Users\UserId\AppData\Roaming\cs_tools\.cs_tools\Lib\site-packages\cs_tools\api\_client.py:88 in request                       │
│                                                                                                                                   │
│    85 │   │   # DEV NOTE: @boonhapus, 2024/02/15                                                                                  │
│    86 │   │   # If we want to make the CS Tools HTTP client asyncio-friendly, all we need to d                                    │
│    87 │   │   # request methods for async counterparts.                                                                           │
│ ❱  88 │   │   return self._session.request(method, url, **kwargs)                                                                 │
│    89 │                                                                                                                           │
│    90 │   def __after_response__(self, response: httpx.Response) -> None:                                                         │
│    91 │   │   """                                                                                                                 │
│                                                                                                                                   │
│ C:\Users\UserId\AppData\Roaming\cs_tools\.cs_tools\Lib\site-packages\httpx\_client.py:827 in request                             │
│                                                                                                                                   │
│    824 │   │   │   timeout=timeout,                                                                                               │
│    825 │   │   │   extensions=extensions,                                                                                         │
│    826 │   │   )                                                                                                                  │
│ ❱  827 │   │   return self.send(request, auth=auth, follow_redirects=follow_redirects)                                            │
│    828 │                                                                                                                          │
│    829 │   @contextmanager                                                                                                        │
│    830 │   def stream(                                                                                                            │
│                                                                                                                                   │
│                                                      ... 2 frames hidden ...                                                      │
│                                                                                                                                   │
│ C:\Users\UserId\AppData\Roaming\cs_tools\.cs_tools\Lib\site-packages\httpx\_client.py:979 in _send_handling_redirects            │
│                                                                                                                                   │
│    976 │   │   │   for hook in self._event_hooks["request"]:                                                                      │
│    977 │   │   │   │   hook(request)                                                                                              │
│    978 │   │   │                                                                                                                  │
│ ❱  979 │   │   │   response = self._send_single_request(request)                                                                  │
│    980 │   │   │   try:                                                                                                           │
│    981 │   │   │   │   for hook in self._event_hooks["response"]:                                                                 │
│    982 │   │   │   │   │   hook(response)                                                                                         │
│                                                                                                                                   │
│ C:\Users\UserId\AppData\Roaming\cs_tools\.cs_tools\Lib\site-packages\httpx\_client.py:1015 in _send_single_request               │
│                                                                                                                                   │
│   1012 │   │   │   )                                                                                                              │
│   1013 │   │                                                                                                                      │
│   1014 │   │   with request_context(request=request):                                                                             │
│ ❱ 1015 │   │   │   response = transport.handle_request(request)                                                                   │
│   1016 │   │                                                                                                                      │
│   1017 │   │   assert isinstance(response.stream, SyncByteStream)                                                                 │
│   1018                                                                                                                            │
│                                                                                                                                   │
│ C:\Users\UserId\AppData\Roaming\cs_tools\.cs_tools\Lib\site-packages\httpx\_transports\default.py:232 in handle_request          │
│                                                                                                                                   │
│   229 │   │   │   content=request.stream,                                                                                         │
│   230 │   │   │   extensions=request.extensions,                                                                                  │
│   231 │   │   )                                                                                                                   │
│ ❱ 232 │   │   with map_httpcore_exceptions():                                                                                     │
│   233 │   │   │   resp = self._pool.handle_request(req)                                                                           │
│   234 │   │                                                                                                                       │
│   235 │   │   assert isinstance(resp.stream, typing.Iterable)                                                                     │
│                                                                                                                                   │
│ C:\Users\UserId\AppData\Local\Programs\Python\Python312\Lib\contextlib.py:158 in __exit__                                        │
│                                                                                                                                   │
│ C:\Users\UserId\AppData\Roaming\cs_tools\.cs_tools\Lib\site-packages\httpx\_transports\default.py:86 in map_httpcore_exceptions  │
│                                                                                                                                   │
│    83 │   │   │   raise                                                                                                           │
│    84 │   │                                                                                                                       │
│    85 │   │   message = str(exc)                                                                                                  │
│ ❱  86 │   │   raise mapped_exc(message) from exc                                                                                  │
│    87                                                                                                                             │
│    88                                                                                                                             │
│    89 HTTPCORE_EXC_MAP = {                                                                                                        │
╰───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
RemoteProtocolError: Server disconnected without sending a response.


                          ╭───────────────────── 🤷  This is an unhandled error!  🤷 ─────────────────────╮
                          │                                                                               │
                          │ If you encounter this message more than once, please help by letting us know! │
                          │                                                                               │
                          │           GitHub: https://github.com/thoughtspot/cs_tools/issues/new/choose   │
                          │                                                                               │
                          ╰──────────── Run cs_tools logs report to send us your last error. ─────────────╯

**Reply by boonhapus** [reply]
hey @apagapulan-cwt , are you able to run the **[users/search API](https://developers.thoughtspot.com/docs/restV2-playground?apiResourceId=http%2Fapi-endpoints%2Fusers%2Fsearch-users)** in your ThoughtSpot cluster at all? Please try it once without any parameters and let me know the result.

The error message is telling us that ThoughtSpot terminated the connection with CS Tools, so I am thinking there may be some issue there.

**Comment by apagapulan-cwt** [comment]
Hello, yup the users/search api works fine when i tested it using postman
